### PR TITLE
MTV-1352: Documentation of MTV performance 2.6

### PR DESCRIPTION
**Pull request by anarnold97**
### JIRA

* [MTV-1352](https://issues.redhat.com/browse/MTV-1352)

### PREVIEW

* [MTV performance recommendations](https://anarnold97.github.io/MTA-Preview/MTV/mtv-performance.html#mtv-performance-recommendation_mtv)
* [MTV performance addendum](https://anarnold97.github.io/MTA-Preview/MTV/mtv-performance.html#mtv-performance-addendum_mtv)

### CODE REVIEW

**Comment by mordechai** [comment]
Following comments for: https://anarnold97.github.io/MTA-Preview/MTV/mtv-performance.html#mtv-performance-recommendation_mtv

**Suggestion 1)**
Line currently reads 'When more ESXi servers are used, high datastore read rates are possible.' 
I think could be reworded to 'When _multiple_ ESXi servers are used, _higher_ datastore read rates are possible.

**Suggestion 2)**
Line currently reads " Testing showed that when transferring more than 10 VMs with both BIOS and host power management set accordingly, dataset read rates between tested VMs increased by 15 MiB."
I think could be reworded to " Testing showed that when transferring more than 10 VMs with both BIOS and host power management set accordingly, migrations had an increase of 15 MiB in the average datastore read rate."

**Suggestion 3)**
Line current reads " Maximum migration transfer rate of a single disk: 162 MiB/second (1.5 Tb case / time using utilized data) "
I think could be reworded to: "Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5T of non-zero data) "

**Suggestion 4)**
Line reads: " In MTV 2.6, MAX_VM_INFLIGHT reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot.  The total set by MAX_VM_INFLIGHT is used to utilize one slot.

Example

    MAX_VM_INFLIGHT = 20 and 2 ESXi hosts defined in the provider mean each host can transfer 20 VMs. "
    
  I think could be reworded to:    " In MTV 2.6, MAX_VM_INFLIGHT reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm.  The total set by MAX_VM_INFLIGHT is used to indicate how many concurrent VM tranfers per ESXi host is allowed.

Example

    MAX_VM_INFLIGHT = 20 used with 2 ESXi hosts defined in the provider implies that each host can transfer 20 VMs. "

**Suggestion 5)**
In section 3.4.5 
Line reads: "The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to openshift occurs."
I think the o should be capitalized for Openshift "The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to _O_penshift occurs."  

In https://anarnold97.github.io/MTA-Preview/MTV/mtv-performance.html#mtv-performance-addendum_mtv   

**Suggestion 6)**

There are various references to the private network such as:
172.x Network or  172.16.x.x Network  I think these references should just be replaced with Private Network
with a foot note that explains Private Network is a non -Management network .

**Comment by fabiand** [comment]
Can we move this into it's dedicated section, aka level it up?

Mostly to avoid deep nesting.

To me it could be a sectoin just before `Troubleshooting?

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+** For example, 200 to 325 MiB/s was the average network transfer rate from the `vmnic` for each ESXi host associated with transferring data to the OCP interface.
+
+[id="mtv-performance-datastore-read_{context}"]
+== Ensure fast datastore read speeds to ensure efficient and performant migrations.
+
+
+Datastores read rates impact the total transfer times, so it is essential to ensure fast reads are possible from the ESXi datastore to the ESXi host.  
+
+
+Example in numbers: 200 to 300 MiB/s was the average read rate for both vSphere and ESXi endpoints for a single ESXi server. When more ESXi servers are used, high datastore read rates are possible.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:35]:
```suggestion
Example in numbers: 200 to 300 MiB/s was the average read rate for both vSphere and ESXi endpoints for a single ESXi server. When multiple ESXi servers are used, higher datastore read rates are possible.
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+* 80 single disk VMs, containing 35 GiB of data each, using a single host took 2 hours and 43 minutes, with a migration rate of 294 MiB/s.
+
+* 80 single disk VMs, containing 35 GiB of data each, using 8 ESXi hosts took 41 minutes, with a migration rate of 1,173 MiB/s.
+
+[id="mtv-performance-bios-profile_{context}"]
+== Set ESXi hosts BIOS profile and ESXi Host Power Management for High Performance
+
+Where possible, ensure that hosts used to perform migrations are set with BIOS' profiles related to maximum performance.  Hosts which use Host Power Management controlled within vSphere should check that ‘High Performance’ is set.
+
+Testing showed that when transferring more than 10 VMs with both BIOS and host power management set accordingly, dataset read rates between tested VMs increased by 15 MiB.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:111]:
```suggestion
Testing showed that when transferring more than 10 VMs with both BIOS and host power management set accordingly, migrations had an increase of 15 MiB in the average datastore read rate.
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+* Maximum number of ESXi hosts tested: 8
+* Maximum number of VMs in a single migration plan: 500
+* Maximum number of VMs migrated: 5000
+* Maximum number of migration plans performed concurrently: 40
+* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
+* Maximum number of disks on a single VM migrated: 50
+* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
+* Highest observed multi-datastore read rate using two ESXi servers and two datastores: 1,242 MiB/second
+* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second
+* Maximum migration transfer rate of a single disk: 162 MiB/second (1.5 Tb case / time using utilized data)
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:140]:
```suggestion
* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5Tb of non-zero data)
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+Set the `MAX_VM_INFLIGHT MTV` variable to control the maximum number of concurrent VMs transfers allowed for the ESXi host. 
+
+{project-short} allows for concurrency to be controlled using this variable; by default, it is set to 20.
+
+When setting `MAX_VM_INFLIGHT`, consider the number of maximum concurrent VMs transfers are required for ESXi hosts. It is important to consider the type of migration to be transferred concurrently. Warm  migrations, which are defined by migrations of a running VM that will be migrated over a scheduled time.
+
+Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to openshift occurs. 
+
+In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot.  The total set by `MAX_VM_INFLIGHT` is used to utilize one slot.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:86]:
```suggestion
In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm. The total set by `MAX_VM_INFLIGHT` is used to indicate how many concurrent VM tranfers per ESXi host is allowed.
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+[id="mtv-performance-concurrent-disk_{context}"]
+== Control maximum concurrent disk migrations per ESXi host.
+
+Set the `MAX_VM_INFLIGHT MTV` variable to control the maximum number of concurrent VMs transfers allowed for the ESXi host. 
+
+{project-short} allows for concurrency to be controlled using this variable; by default, it is set to 20.
+
+When setting `MAX_VM_INFLIGHT`, consider the number of maximum concurrent VMs transfers are required for ESXi hosts. It is important to consider the type of migration to be transferred concurrently. Warm  migrations, which are defined by migrations of a running VM that will be migrated over a scheduled time.
+
+Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to openshift occurs.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:84]:
```suggestion
Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to {ocp-short} occurs. 
```

**Comment by anarnold97** [comment]
* [Updated preview](https://anarnold97.github.io/MTA-Preview/MTV/mtv-performance.html#mtv-performance-recommendation_mtv)

* [Updated performance addendum](https://anarnold97.github.io/MTA-Preview/MTV/mtv-performance.html#mtv-performance-addendum_mtv)

* Updated taxonomy

image

**Comment by dgur** [comment]
Lines that in the Beginning  
"The purpose of this section is to show the current observed experience in the process of testing the performance and scale of Migration Toolkit for Virtualization (MTV), to migrate VMs from vSphere to OpenShift Virtualization. 

This document also aims to share key findings of what was observed through testing, and the summarized results of those tests, as shown by recommendations."

- those are from the original full performance document,   I  this official doc we don't show observed  experience  and test results -  I believe, We should state something  like this :
"The purpose of this section is to share recommendations for performant VMs migrations using MTV, based on findings observed through testing"

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 6]
```diff
@@ -0,0 +1,142 @@
+// Module included in the following assemblies:
+//
+// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc
+
+:_content-type: CONCEPT
+[id="mtv-performance-recommendation_{context}"]
+= {project-short} performance recommendations
+
+The purpose of this section is to show the current observed experience in the process of testing the performance and scale of {project-first}, to migrate VMs from vSphere to {virt}.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:9]:
```suggestion
The purpose of this section is to share recommendations for efficient and effective migration of virtual machines (VMs) using {project-first}, based on findings observed through testing.
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm. The total set by `MAX_VM_INFLIGHT` is used to indicate how many concurrent VM tranfers per ESXi host is allowed.
+
+.Example
+
+* `MAX_VM_INFLIGHT = 20` and 2 ESXi hosts defined in the provider mean each host can transfer 20 VMs.
+
+[id="mtv-performance-multiple-vms-concurrently_{context}"]
+== Migrations complete faster when migrating multiple VMs concurrently
+
+Migrations complete faster when migrating multiple VMs concurrently.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:94]:
```suggestion
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to {ocp-short} occurs. 
+
+In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm. The total set by `MAX_VM_INFLIGHT` is used to indicate how many concurrent VM tranfers per ESXi host is allowed.
+
+.Example
+
+* `MAX_VM_INFLIGHT = 20` and 2 ESXi hosts defined in the provider mean each host can transfer 20 VMs.
+
+[id="mtv-performance-multiple-vms-concurrently_{context}"]
+== Migrations complete faster when migrating multiple VMs concurrently
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:92]:
```suggestion
== Migrations are completed faster when migrating multiple VMs concurrently
```

File: documentation/modules/mtv-performance-addendum.adoc [hunk 9]
```diff
@@ -0,0 +1,157 @@
+// Module included in the following assemblies:
+//
+// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc
+
+:_content-type: CONCEPT
+[id="mtv-performance-addendum_{context}"]
+= {project-short} performance addendum
+
```
Comment (fabiand) [inline documentation/modules/mtv-performance-addendum.adoc:8]: > The data provided here was gathered in Red Hat Labs and  is provided for reference only. Overall these numbers can be considered to show the best-case scenarios.
> Observed migration performance can differ from these results and depends on a number of factors.

@mordechai we should include something like this.

File: documentation/modules/mtv-performance-addendum.adoc [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+[id="mtv-single-esxi-performance_{context}"]
+== Single ESXi performance
+
+Test migration using the same ESXi host.
+
+In each iteration, the total VMs are increased, to display the impact of concurrent migration on the duration.
+
+The results show that migration time is linear when increasing the total VMs (Utilization 70%).
+
+The optimal number of VMs per ESXi is 10.
```
Comment (fabiand) [inline documentation/modules/mtv-performance-addendum.adoc:18]: Should this be moved into the single ESXi section?

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+//
+// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc
+
+:_content-type: CONCEPT
+[id="mtv-performance-recommendation_{context}"]
+= {project-short} performance recommendations
+
+The purpose of this section is to share recommendations for efficient and effective migration of virtual machines (VMs) using {project-first}, based on findings observed through testing.
+
+This document also aims to share key findings of what was observed through testing, and the summarized results of those tests, as shown by recommendations.
```
Comment (fabiand) [inline documentation/modules/mtv-performance-recommendation.adoc:11]: A similar disclaimer is needed for the abendum as well.

**Review by fabiand — COMMENTED** [review]
@mordechai please review

@anarnold97 please include the disclaimer in the abendum as well.

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+* Maximum number of ESXi hosts tested: 8
+* Maximum number of VMs in a single migration plan: 500
+* Maximum number of VMs migrated in a single test: 5000
+* Maximum number of migration plans performed concurrently: 40
+* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
+* Maximum number of disks on a single VM migrated: 50
+* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
+* Highest observed multi-datastore read rate using eight ESXi servers and two datastores: 1,242 MiB/second
+* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second
+* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5Tb case/time using utilized data)
```
Comment (dgur) [inline documentation/modules/mtv-performance-recommendation.adoc:170]: rate observed when transferring 1.5Tb case/time using utilized data - ether remove ether change to "nonconcurrent migration of  1.5Tb utilized data)

File: documentation/modules/mtv-performance-addendum.adoc [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+[id="mtv-esxi-performance_{context}"]
+== ESXi performance
+
+.Single ESXi performance
+
+Test migration using the same ESXi host.
+
+In each iteration, the total VMs are increased, to display the impact of concurrent migration on the duration.
+
+The results show that migration time is linear when increasing the total VMs (Utilization 70%).
```
Comment (dgur) [inline documentation/modules/mtv-performance-addendum.adoc:20]: **(50 GB disk**, Utilization 70%)
We better say how much data is moved

File: documentation/modules/mtv-performance-addendum.adoc (hunk 10 above)
Comment (dgur) [reply documentation/modules/mtv-performance-addendum.adoc:22]: We write the same in the recommendation of Single ESX concurency

File: documentation/modules/mtv-performance-addendum.adoc [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+|2.6
+|7.0.3
+|100
+|cold
+|3:25:15
+|===
+
+.Multi ESXi hosts and single data store
+
+In each iteration, the number of ESXi hosts were increased, to show that increasing the number of ESXi hosts improves the migration time.
```
Comment (dgur) [inline documentation/modules/mtv-performance-addendum.adoc:87]: Add the size of VMs, it is important (50 GB Disk, 70% utilized)

File: documentation/modules/mtv-performance-addendum.adoc [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+|===
+| Test Case Description
+| MTV
+| VDDK
+| max_vm inflight
+| Migration Type
+| Total Duration
+
+
+|cold migration , 10 VMs , ESX 02-03 , Private Network footnoteref:[note1,Private Network refers to a non -Management network]
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-addendum.adoc:35]:
```suggestion
|cold migration , 10 VMs , Single ESX , Private Network footnoteref:[note1,Private Network refers to a non -Management network]
```

File: documentation/modules/mtv-performance-addendum.adoc [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+|2.6
+|7.0.3
+|100
+|cold
+|3:25:15
+|===
+
+.Multi ESXi hosts and single data store
+
+In each iteration, the number of ESXi hosts were increased, to show that increasing the number of ESXi hosts improves the migration time.
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-addendum.adoc:87]:
```suggestion
In each iteration, the number of ESXi hosts were increased, to show that increasing the number of ESXi hosts improves the migration time (50 GiB disk, Utilization 70%).
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+== Migrations are completed faster when migrating multiple VMs concurrently
+
+
+When multiple VMs from a specific ESXi host are to be migrated, starting concurrent migrations for multiple VMs leads to faster migration times. 
+
+Testing demonstrated that migrating 10 VMs (each containing 35 GiB of data, with a total size of 50 GiB) from a single host is significantly faster than migrating the same number of VMs sequentially, one after another. 
+
+It is possible to increase concurrent migration to more than 10 virtual machines from a single host, but it does not show a significant improvement. 
+
+.Example
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc:101]:
```suggestion
.Examples
```

File: documentation/modules/mtv-performance-addendum.adoc [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+|0:21:30
+
+|cold migration , 10 VMs , ESXi 01-03 , Private Network footnoteref:[note2,Private Network refers to a non-Management network]
+|2.6
+|7.0.3
+|20
+|cold
+|0:21:20
+
+|cold migration , 10 VMs , 1 ESXi(0203) , Default Network
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-addendum.adoc:155]:
```suggestion
|cold migration , 10 VMs , Single ESXi , Default Network
```

File: documentation/modules/mtv-performance-addendum.adoc [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+|Total Duration
+
+|cold migration , 10 VMs , Single ESXi , MGMT Network
+|2.6
+|7.0.3
+|100
+|cold
+|0:21:30
+
+|cold migration , 10 VMs , ESXi 01-03 , Private Network footnoteref:[note2,Private Network refers to a non-Management network]
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-addendum.adoc:148]:
```suggestion
|cold migration , 10 VMs , Single ESXi , Private Network footnoteref:[note2,Private Network refers to a non-Management network]
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+* Maximum number of ESXi hosts tested: 8
+* Maximum number of VMs in a single migration plan: 500
+* Maximum number of VMs migrated in a single test: 5000
+* Maximum number of migration plans performed concurrently: 40
+* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
+* Maximum number of disks on a single VM migrated: 50
+* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
+* Highest observed multi-datastore read rate using eight ESXi servers and two datastores: 1,242 MiB/second
+* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second
+* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5Tb case/time using utilized data)
```
Comment (anarnold97) [inline documentation/modules/mtv-performance-recommendation.adoc]:
```suggestion
* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring nonconcurrent migration of 1.5 Tb utilized data)
```

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+
+When multiple VMs from a specific ESXi host are to be migrated, starting concurrent migrations for multiple VMs leads to faster migration times. 
+
+Testing demonstrated that migrating 10 VMs (each containing 35 GiB of data, with a total size of 50 GiB) from a single host is significantly faster than migrating the same number of VMs sequentially, one after another. 
+
+It is possible to increase concurrent migration to more than 10 virtual machines from a single host, but it does not show a significant improvement. 
+
+.Example
+
+* 1 single disk VMs took 6 minutes, with migration rate of 100 MiB/s
```
[the remark this answers is not in the document]
Comment (anarnold97) [reply documentation/modules/mtv-performance-recommendation.adoc:112]: image

File: documentation/modules/mtv-performance-recommendation.adoc [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+* Maximum number of ESXi hosts tested: 8
+* Maximum number of VMs in a single migration plan: 500
+* Maximum number of VMs migrated in a single test: 5000
+* Maximum number of migration plans performed concurrently: 40
+* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
+* Maximum number of disks on a single VM migrated: 50
+* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
+* Highest observed multi-datastore read rate using eight ESXi servers and two datastores: 1,242 MiB/second
+* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second
+* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring nonconcurrent migration of 1.5 Tb utilized data)
```
Comment (dgur) [inline documentation/modules/mtv-performance-recommendation.adoc:173]: Can you bold those below  3 as well 
-maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring nonconcurrent migration of 1.5 Tb utilized data)

 Maximum cold migration transfer rate of the multiple VMs (single disk) from a single ESXi host: 294 MiB/s (concurrent migration of 30 VMs, 35/50 GiB used, from Single ESXi)
* Maximum cold migration transfer rate of the multiple VMs (single disk) from multiple ESXi hosts: 1173MB/s (concurrent migration of 80 VMs, 35/50 GiB used, from 8 ESXi servers,

**Comment by anarnold97** [comment]
|    |      |
| ---- | ---- |
| image | image |

Updated

**Pull request merged**